### PR TITLE
Avoid sharing ReferenceManager across script compilations that do not share previous submission.

### PR DESCRIPTION
**Pull request by tmat**
Fixes crashes in integration tests.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -609,7 +610,13 @@ public CSharpCompilation WithScriptCompilationInfo(CSharpScriptCompilationInfo?
...
             }
 
-            // Reference binding doesn't depend on previous submission so we can reuse it.
+            // Metadata references are inherited from the previous submission,
+            // so we can only reuse the manager if we can guarantee that these references are the same.
+            // Check if the previous script compilation doesn't change. 
+
+            // TODO: Consider comparing the metadata references if they have been bound already.
+            // https://github.com/dotnet/roslyn/issues/43397
+            bool reuseReferenceManager = ReferenceEquals(ScriptCompilationInfo?.PreviousScriptCompilation, info?.PreviousScriptCompilation);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:619]: >bool reuseReferenceManager = ReferenceEquals(ScriptCompilationInfo?.PreviousScriptCompilation, info?.PreviousScriptCompilation); [](start = 12, length = 128)

It looks like we have similar API with the same issue on VB side.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by AlekseyTs** [comment]
@tmat It looks like CI is failing. Could that be a legitimate failure?

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 2]
```diff
@@ -661,7 +661,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
             End If
 
-            ' Reference binding doesn't depend on previous submission so we can reuse it.
+            ' Metadata references are inherited from the previous submission,
+            ' so we can only reuse the manager if we can guarantee that these references are the same.
+            ' Check if the previous script compilation doesn't change. 
+
+            ' TODO Consider comparing the metadata references if they have been bound already.
+            ' https://github.com/dotnet/roslyn/issues/43397
+            Dim reuseReferenceManager = ReferenceEquals(ScriptCompilationInfo?.PreviousScriptCompilation, info?.PreviousScriptCompilation)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:670]: >ReferenceEquals(ScriptCompilationInfo?.PreviousScriptCompilation, info?.PreviousScriptCompilation) [](start = 40, length = 98)

VB has ```Is``` operator to compare references, please use it.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

**Comment by CyrusNajmabadi** [comment]
@tmat can this go in?  Thanks!

**Comment by tmat** [comment]
@CyrusNajmabadi Merged

**Pull request merged**